### DISCUSSION TITLE
Is there any documentation on the VIA

### DISCUSSION THREAD

**Discussion by midicdj1000** [picocomputer/community#33] [Q&A]
Not sure if I missed it, but I would like to add a basic atari joystick or perhaps a retro keyboard.
Guessing it's very must like on the C64

**Comment by brentward** [comment]
It is much like a C64 although the C64 used a 6526 CIA instead of the 6522 VIA. Nevertheless it is very similar and I used code that was intended to read a SNES controller on a C64 user port with very little modification.

I hooked up 2 SNES controllers to the VIA. If you want an example of how to interface with the VIA my code both reads and writes to the pins;  you should be able to adapt it to an Atari joystick which is simpler because it uses 1 pin per button I instead of latch and clock like the NES/SNES.

Here is my code:
https://github.com/brentward/rp6502_hello_world

Here is the documentation for the VIA:
https://www.westerndesigncenter.com/wdc/documentation/w65c22.pdf

**Reply by midicdj1000** [reply]
i should have said, is there a register or address I use to talk to it, but now I have some time I will take a closer look and I also made a snes adapter so I can try your code.

**Comment by brentward** [comment]
Here is some basic code to read an atari style joystick.
https://github.com/brentward/rp6502-atari-joystick
This uses PA0 to PA4 for the input pins. I detailed which input goes to which pin in the comments in the assembly.
For it to work you will need pullup resistors between pin 1 and each input pin to pull each input up to 3.3v since the joystick works by grounding each input as it is activated.
I tested on a breadboard that it basically works but I manually hooked up switches to each input pin instead of using a real joystick.
The pullups I used were 10k ohm.

To answer your question about registers, in the rp6502.inc assembly include file there are registers to use in assembly. VIA_DDRA controls the direction of port A and VIA_PA1 is the port A IO register.

As far as using the VIA registers in C directly I dont' think that is included but CC65 does have a header for a 6522 so you could set it up. Try:
```
#include <_6522.h>
#define VIA    (*(volatile struct __6522 *)0xFFD0)
```

Maybe this will be added to the rp6502.h  at some point?

**Reply by brentward** [reply]
After looking at a couple more resources some of the Atari style joysticks will have pin 7 go to +5v and some of them have internal pull up resistors. I would recommend hooking up power and ground to your actual joystick and check if the input pins are already high and if so you could potentially leave out the pull up resistors. You should use +3.3v for power instead of +5v because putting 5v on PA pins is out of spec when VDD is +3.3v.

**Reply by midicdj1000** [reply]
my joystick is a basic 2600 one so nothing inside apart from the dome switches, I never thought about resistors and don't have anything at home so will have to try this tomorrow.

Many thanks

**Reply by brentward** [reply]
With that style of controller hooking up pin 7 to positive voltage shouldn't be necessary, but it will require pull up resistors. For resistor values 10 kohm worked fine for me. From what I read anything 1k or up should work and 4.7 kohm is a good place to start. If you are using the 65c22s that was in the parts list you should be able to go into the megaohm range and it will still work because it has a very low leakage current.

**Comment by BruceMcF** [comment]
I don't have a board -- if you get a built board, how do you get to the VIA pins?

**Reply by rumbledethumps** [reply]
J1 also marked GPIO.
https://www.amazon.com/uxcell-Connector-Ribbon-Female-Length/dp/B07FLRW1ZL

**Reply by BruceMcF** [reply]
Oh! it's screened on the front of the board, like the solder video 2:25. I was looking at the back at, eg, 2:19.

How much room is there in between the header and the acrylic "case", if using the size standoffs recommended in the "no solder / no breadboard" video?

NB. This topic was also raised at 6502.org. There I noted that for my purposes, the best solution is either a right angle block pin header or drilling jumper holes in the acrylic.
